share/getter.go

```diff
@@ -47,25 +47,25 @@ type NamespacedRow struct {
 
 // Verify validates NamespacedShares by checking every row with nmt inclusion proof.
 func (ns NamespacedShares) Verify(root *Root, nID namespace.ID) error {
-	originalRoots := make([][]byte, 0)
-	for _, row := range root.RowsRoots {
-		if !nID.Less(nmt.MinNamespace(row, nID.Size())) && nID.LessOrEqual(nmt.MaxNamespace(row, nID.Size())) {
-			originalRoots = append(originalRoots, row)
+	for _, row := range ns {
+		err := row.verifyAny(root, nID)
+		if err != nil {
+			return err
 		}
 	}
+	return nil
+}
 
-	if len(originalRoots) != len(ns) {
-		return fmt.Errorf("amount of rows differs between root and namespace shares: expected %d, got %d",
-			len(originalRoots), len(ns))
-	}
-
-	for i, row := range ns {
-		// verify row data against row hash from original root
-		if !row.verify(originalRoots[i], nID) {
-			return fmt.Errorf("row verification failed: row %d doesn't match original root: %s", i, root.Hash())
+// verifyAny validates the NMT proof inside NamespacedRow by checking if the NMT
+// inclusion proof is satisfied for any rowRoot from the data availability
+// header.
+func (row *NamespacedRow) verifyAny(root *Root, nID namespace.ID) error {
+	for _, root := range root.RowsRoots {
+		if row.verify(root, nID) {
+			return nil
 		}
 	}
-	return nil
+	return fmt.Errorf("no NMT inclusion proof verified successfully for namespace %v", nID.String())
```

Comment:
@Wondertan @evan-forbes what are your thoughts on this approach? The previous implementation depended on the fact that the row roots were constructed with the property: `isMaxNamespace=true` so that it could efficiently generate a list of `originalRoots` and then use those roots to validate the NMT proofs of each NamespacedRow.

With the new approach, we remove the dependency on the `isMaxNamespace=true` property at the computational cost of attempting to validate one NMT proof per row root in the DAH.

Reply:
An alternative is to change how parity namespace is constructed in celestia-app so that we can still use the `ignoreMaxNamespace=true` feature of the NMT. I'll explore that now in case there are concerns on computational complexity of attempting verify NMT proofs w/ multiple roots

Reply:
opted to not go with this approach so resolving.